Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -1154,8 +1219,8 @@ protected class ClassWriter : IClassWriter {
       return block;
     }
 
-    static string Keywords(bool isPublic = false, bool isStatic = false, bool isExtern = false) {
-      return (isPublic ? "public " : "") + (isStatic ? "static " : "") + (isExtern ? "extern " : "");
+    static string Keywords(bool isPublic = false, bool isStatic = false, bool isExtern = false, bool isVirtual = true) {
+      return (isPublic ? "public " : "") + (isStatic ? "static " : "") + (isExtern ? "extern " : "") + (isVirtual && !isStatic && isPublic ? "virtual " : "");
```

Comment:
Does `isVirtual == true && isStatic == true` ever occur? Consider making this method less smart if these inconsistencies never occur.

Reply:
I rewrote this to refer directly to `DafnyOptions.O.CompileMocks`. The check for `isStatic` seems to be necessary for the generated C# code to compile.